src/chat/ChatScreen.js

```diff
@@ -89,9 +91,20 @@ const useMessagesWithFetch = args => {
   // dropped it.)
   React.useEffect(scheduleFetch, [eventQueueId]);
 
-  // On first mount, fetch. Also unset `shouldFetchWhenNextFocused.current`
-  // that was set in the previous `useEffect`, so the fetch below doesn't
-  // also fire.
+  // If we stop having any data at all about the messages in this narrow --
+  // we don't know any, and nor do we know if there are some -- then
+  // schedule a fetch.  (As long as we have *some* messages, we'll show a
+  // proper MessageList, and scroll events can cause us to fetch newer or
+  // older messages.  But with none, we'll show NoMessages.)
+  //
+  // TODO: We may still briefly show NoMessages; this render will have
+  //   isFetching false, even though the fetch effect will cause a rerender
+  //   with isFetching true.  It'd be nice to avoid that.
+  const nothingKnown = messages.length === 0 && !caughtUp.older && !caughtUp.newer;
+  useEdgeTriggeredEffect(scheduleFetch, nothingKnown, true);
```

Comment:
Yeah, so ideally I think these two (on first mount, and edge-triggered on `nothingKnown`) would be consolidated. In this PR it comes out this way because I (a) didn't want to disturb what happens on first mount (because that currently works at least OK, and this is a tricky area to get right), and (b) wanted something relatively simple and narrowly-targeted for this additional case, to avoid inadvertently causing disruption in cases where things had been working smoothly.

Point (a) meant the first-mount fetch would stay broad, and point (b) meant the new edge-triggered fetch would be narrow, so that blocks combining them.

The existing thing isn't actually an unconditional fetch, even though it looks that way from this code. Over in `fetchActions.js`, it ends up conditioned on this before actually doing any fetching:
```js
  // Ideally this would detect whether, even if we don't have *all* the
  // messages in the narrow, we have enough of them around the anchor
  // to show a message list already.  For now it's simple and cautious.
  const caughtUp = getCaughtUpForNarrow(state, narrow);
  return !(caughtUp.newer && caughtUp.older);
```
So if we know we've fetched the whole narrow already, we don't re-fetch. Otherwise we fetch. And the comment speaks to your question about if we already have a screenful or so of messages.

One reason I wanted (b), to keep the new thing narrowly targeted, is that when we do one of these fetches it replaces whatever else we knew in that narrow: see the logic around `replaceExisting` in narrowsReducer, and note that these fetches have anchor FIRST_UNREAD_ANCHOR. So if the new thing were to have a condition that sometimes triggered when the messages around the first unread didn't include the messages you'd been looking at, then that could be disruptive.

---

The reason I pass `true` for `includeStart` here is just that within this condition's own logic, if we know nothing about the narrow's messages then that's just as good a reason to fetch following first mount as it is later. It is redundant with the first-mount fetch below, but harmlessly so -- just like the `eventQueueId` line -- so it's free to follow that internal logic and not specially avoid fetching in that case.